detection/detect_slow_nodes.py

```diff
@@ -229,7 +374,146 @@ def __sortNodesByExecutionTime(self, nodes: list):
         # return sorted(nodes, key=lambda n: self.__getNumberOfSlowRanksOnNode(n))
         return sorted(node_times, key=lambda t: node_times[t])
 
-    def __findHighOutliers(self, data):
+    def __clusterTimes(self, data):
+        print(f"Beginning clustering for {len(data)} rank total times...")
+        from sklearn.cluster import MeanShift
+
+        data = np.array(data)
+
+        ms = MeanShift(n_jobs=-1).fit(data.reshape(-1, 1))
+        clusters = ms.predict(data.reshape(-1, 1))
+
+
+        cluster_to_times = {}
+        cluster_to_ranks = {}
+
+        for rank, (time, cluster) in enumerate(zip(data, clusters)):
+            if cluster not in cluster_to_times:
+                cluster_to_times[cluster] = []
+                cluster_to_ranks[cluster] = []
+            cluster_to_times[cluster].append(time)
+            cluster_to_ranks[cluster].append(rank)
+
+        cluster_centers = dict(zip(sorted(cluster_to_times.keys()), list(ms.cluster_centers_.reshape(1, -1)[0])))
+
+        representative_cluster = max(cluster_to_times.items(), key=lambda v: len(v[1]))[0]
+        representative_center = cluster_centers[representative_cluster]
+        threshold = representative_center + 3 * np.std(cluster_to_times[representative_cluster])
```

Comment:
Should be make this `3` a parameter instead of a constant used directly in the code?